packages/components/select/src/select.js

```diff
@@ -0,0 +1,328 @@
+		window.addEventListener('mousedown', this.clickOutsideListener);
+	}
+
+	disconnectedCallback() {
+		window.removeEventListener('resize', this.resizeListener);
+		window.removeEventListener('mousedown', this.clickOutsideListener);
+	}
+
+	update(changedProperties) {
+		if (changedProperties.has('selected')) {
+			// initialise a current selection with selected items.
+			this.currentSelection = [...this.selected];
+		}
+		if (changedProperties.has('opened')) {
+			// clean filter values on every open/closed state change.
+			this.inputValue = '';
+			this.filterValue = '';
+			this.showSelectedOnly = false;
+		}
+		super.update(changedProperties);
+	}
+
+	attributeChangedCallback(name, oldVal, newVal) {
+		super.attributeChangedCallback(name, oldVal, newVal);
+		switch (name) {
+			case 'opened':
+				if (newVal === null) {
+					// discard changes by assigning previously selected items to current selection
+					this.currentSelection = [...this.selected];
+					this.dirty = false;
+				}
+				if (newVal !== null) {
+					window.customElements.whenDefined('ts-select').then(() => {
+						this.updatePosition();
+					});
+				}
+				break;
+
+			default:
+			// do nothing
+		}
+	}
+
+	/**
+	 * @private
+	 * @param {Event} event
+	 */
+	handleOutsideClick(event) {
+		if (!this.opened) {
+			return;
+		}
+		if (!('composedPath' in event)) {
+			return;
+		}
+		const paths = event.composedPath();
+		const container = this.shadowRoot.getElementById('container');
+		if (paths.indexOf(container) === -1) {
+			this.opened = false;
+		}
+	}
+
+	/**
+	 * Calculates the size and position of the dropdown part.
+	 * @private
+	 */
+	updatePosition() {
+		if (!this.opened) {
+			return;
+		}
+		requestAnimationFrame(() => {
+			const selectInput = this.shadowRoot.getElementById('selectInput');
+			const dropdown = this.shadowRoot.getElementById('dropdown');
+			const listContainer = this.shadowRoot.getElementById('listContainer');
+			if (!dropdown) {
+				// it could happen when element is just initialised but not rendered yet.
+				return;
+			}
+			const { height, width, top, bottom } = selectInput.getBoundingClientRect();
+			const windowHeight = window.innerHeight;
+			const position = { top: 'auto', bottom: 'auto', width: width + 'px' };
+			let maxHeight = 0;
+			const paddingTop = top - 20;
+			const paddingBottom = windowHeight - bottom - 20;
+			if (paddingBottom < MAX_HEIGHT && paddingBottom < paddingTop) {
+				// When we don't have enough space below and have more space above, we put dropdown above.
+				maxHeight = top > MAX_HEIGHT ? MAX_HEIGHT : paddingTop;
+				position.bottom = height + 10 + 'px';
+			} else {
+				position.top = height + 10 + 'px';
+				maxHeight = paddingBottom > MAX_HEIGHT ? MAX_HEIGHT : paddingBottom;
+			}
+
+			maxHeight -= APPLY_CONTAINER_HEIGHT;
+			maxHeight += 'px';
+
+			Object.assign(dropdown.style, position);
+			Object.assign(listContainer.style, { maxHeight });
```

Comment:
I couldn't figure it out why but the max-height doesn't match the design as I can see. The design itself has inconsistency, the top part says it has max-height of 600px but in the examples when there's a scrollbar, you can see its height is less than 400px. We should check this with @mihneaz

Reply:
I did not notice the section above the `Select component` in Figma, so I choose this `400px` just as a reasonable value for me. Changed to `600px`.